backend/cmm_helpers.ml

```diff
@@ -2333,34 +2355,55 @@ let unaligned_set_16 ptr idx newval dbg =
     let v2 = Cop (Cand, [newval; cconst_int 0xFF], dbg) in
     let b1, b2 = if Arch.big_endian then v1, v2 else v2, v1 in
     Csequence
-      ( Cop (Cstore (Byte_unsigned, Assignment), [add_int ptr idx dbg; b1], dbg),
+      ( Cop
+          ( Cstore (Byte_unsigned, Assignment),
+            [add_int_ptr ~ptr_out_of_heap ptr idx dbg; b1],
+            dbg ),
         Cop
           ( Cstore (Byte_unsigned, Assignment),
-            [add_int (add_int ptr idx dbg) (cconst_int 1) dbg; b2],
+            [ add_int_addr
+                (add_int_ptr ~ptr_out_of_heap ptr idx dbg)
+                (cconst_int 1) dbg;
+              b2 ],
             dbg ) )
 
-let unaligned_load_32 ptr idx dbg =
+let unaligned_load_32 ~ptr_out_of_heap ptr idx dbg =
   if Arch.allow_unaligned_access
-  then Cop (mk_load_mut Thirtytwo_unsigned, [add_int ptr idx dbg], dbg)
+  then
+    Cop
+      ( mk_load_mut Thirtytwo_unsigned,
+        [add_int_ptr ~ptr_out_of_heap ptr idx dbg],
+        dbg )
   else
     let cconst_int i = Cconst_int (i, dbg) in
-    let v1 = Cop (mk_load_mut Byte_unsigned, [add_int ptr idx dbg], dbg) in
+    let v1 =
+      Cop
+        ( mk_load_mut Byte_unsigned,
+          [add_int_ptr ~ptr_out_of_heap ptr idx dbg],
+          dbg )
+    in
     let v2 =
       Cop
         ( mk_load_mut Byte_unsigned,
-          [add_int (add_int ptr idx dbg) (cconst_int 1) dbg],
+          [ add_int_addr
+              (add_int_ptr ~ptr_out_of_heap ptr idx dbg)
+              (cconst_int 1) dbg ],
           dbg )
     in
     let v3 =
       Cop
         ( mk_load_mut Byte_unsigned,
-          [add_int (add_int ptr idx dbg) (cconst_int 2) dbg],
+          [ add_int_addr
+              (add_int_ptr ~ptr_out_of_heap ptr idx dbg)
+              (cconst_int 2) dbg ],
```

Comment:
maybe for a later PR, but maybe we could instead do:
```ocaml
(add_int_ptr ~ptr_out_of_heap ptr (add_int idx 2 dbg) dbg)
```
which would at least optimize the case where `idx` is a constant.

Reply:
Ah there's a mistake here as well, the `add_int_addr` was supposed to be `add_int_ptr`.

Reply:
I've left a CR for the rearrangement though.